### PR TITLE
[BEAM-2915] Add support for handling bag user state to the java-fn-execution library to support runner integration.

### PR DESCRIPTION
**Pull request by lukecwik**
Note that we specifically use an optimization which significantly simplifies the logical stream handling during get/append since we don't need to encode/decode elements to actually find the element boundaries. We just store and replay the chunks the user gave us significantly reducing the amount of pipeline manipulation the needs to happen.

------------------------

Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

It will help us expedite review of your Pull Request if you tag someone (e.g. `@username`) to look at it.

### CODE REVIEW

**Comment by lukecwik** [comment]
CC: @tweise @bsidhom @tgroh

**Comment by lukecwik** [comment]
Run Java PreCommit

File: runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/graph/GreedyPCollectionFusers.java [hunk 1]
```diff
@@ -181,6 +181,10 @@ private static boolean parDoCompatibility(
     // side inputs can be fused with other transforms in the same environment which are not
     // upstream of any of the side inputs.
     return pipeline.getSideInputs(parDo).isEmpty()
+        // Since we lack the ability to mark upstream transforms as key preserving, we
+        // purposefully break fusion here to provide runners the opportunity to insert a
+        // grouping operation
+        && pipeline.getUserStates(parDo).isEmpty()
```
Comment (tweise) [inline runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/graph/GreedyPCollectionFusers.java]: Based on what information would a runner insert a shuffle though? Unless that is available, there is also no benefit breaking the chain here?

File: runners/java-fn-execution/src/main/java/org/apache/beam/runners/fnexecution/state/StateRequestHandlers.java [hunk 2]
```diff
@@ -214,17 +258,22 @@ public static StateRequestHandler forMultimapSideInputHandlerFactory(
...
                 .get(stateKey.getSideInputId());
-        MultimapSideInputHandler<?, ?, ?> handler = cache.computeIfAbsent(
-            sideInputReferenceSpec,
+        MultimapSideInputHandler<Object, Object, BoundedWindow> handler = cache.computeIfAbsent(
+            referenceSpec,
             this::createHandler);
 
+        // TODO: Migrate to using the ByteStringCoder everywhere for user types and values
+        // to reduce the amount of byte array copying.
+        Object key = referenceSpec.keyCoder().decode(stateKey.getKey().newInput());
```
Comment (tweise) [inline runners/java-fn-execution/src/main/java/org/apache/beam/runners/fnexecution/state/StateRequestHandlers.java]: Does this mean only key types with built-in coders are supported? Why do we need to decode the key here?

Comment (lukecwik) [reply runners/java-fn-execution/src/main/java/org/apache/beam/runners/fnexecution/state/StateRequestHandlers.java]: No, all types are supported. We are hiding the fact that we are using the ByteStringCoder internally (hardcoded in ProcessBundleDescriptors#forBagUserState) as an optimization.

This allows us to pass an Object and a Coder without the Runner needing to care how to turn that into bytes giving us the freedom to change which types we pass around without the Runner being the wiser. The TODO is about changing the interface to be tailored towards using ByteString directly on the interface and would be a minor performance improvement for some runners.

**Comment by tweise** [comment]
test this please

**Comment by tweise** [comment]
I tried to build locally and had to make the following change:
```
diff --git a/runners/java-fn-execution/src/test/java/org/apache/beam/runners/fnexecution/control/RemoteExecutionTest.java b/runners/java-fn-execution/src/test/java/org/apache/beam/runners/fnexecution/control/RemoteExecutionTest.java
index ca93d5ecc2..67b2f735c6 100644
--- a/runners/java-fn-execution/src/test/java/org/apache/beam/runners/fnexecution/control/RemoteExecutionTest.java
+++ b/runners/java-fn-execution/src/test/java/org/apache/beam/runners/fnexecution/control/RemoteExecutionTest.java
@@ -45,6 +45,7 @@ import java.util.concurrent.ThreadFactory;
 import org.apache.beam.fn.harness.FnHarness;
 import org.apache.beam.model.fnexecution.v1.BeamFnApi.Target;
 import org.apache.beam.model.pipeline.v1.RunnerApi;
+import org.apache.beam.model.pipeline.v1.RunnerApi.Components;
 import org.apache.beam.runners.core.construction.PipelineTranslation;
 import org.apache.beam.runners.core.construction.graph.ExecutableStage;
 import org.apache.beam.runners.core.construction.graph.FusedPipeline;
```
Unfortunately there seem to be also unrelated test failures.

**Comment by lukecwik** [comment]
I have been working to figure out what it is that is breaking this and it seems to be related to shading beam-model-pipeline with the direct runner.

**Comment by lukecwik** [comment]
I looked at the post shaded beam-model-pipeline proto filedescriptors and it turns out that during the shading process, an over eager string replacement inside a class is corrupting an internal field, in this case it modifies the RunnerApi file descriptor storing `org.apache.beam.model.pipeline.v1.AccumulationMode.Enum` and changing it to `org.apache.beam.repackaged.beam_runners_direct_java.model.pipeline.v1.AccumulationMode.Enum`

This problem exists because the proto package name and the java package name collide and use `org.apache.beam.model.pipeline.v1`

**Comment by lukecwik** [comment]
Filed BEAM-4484 for the shading issue. Will investigate to see if there is a way to keep shading beam-model-pipeline in the direct runner.

**Comment by tweise** [comment]
Thanks, can you rebase this PR to fix the compile error? Since the shading issue is pre-existing in master, I think we can merge this.

**Comment by lukecwik** [comment]
I thought I already did rebase to fix the `Components` issue.

I can't merge because the tests will fail consistently until I solve/work around BEAM-4484

**Comment by lukecwik** [comment]
It turns out that hcatalog tests always fail and are already set to be ignored and the elastic search io tests are flaky. The DirectRunner issue is pre-existing and I'm not making the problem worse here.

Kicked off another test run.

File: runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/graph/GreedyPCollectionFusers.java (hunk 1 above)
Comment (lukecwik) [reply runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/graph/GreedyPCollectionFusers.java]: Runners need to look at the state spec and see if it exists, that would signal insertion of anything which does a partition by key (e.g. a reshuffle). Most runners (maybe all) only support user state via key partitioning.

**Comment by tweise** [comment]
Not clear why some PR pre-commit builds pass and others not. I wasn't able to build master locally. Flaky build?

**Comment by lukecwik** [comment]
It could be flaky, cut bugs for failing tests and assign to authors. Disable the test if it is consistently failing and update the bug saying that it has been disabled.

**Comment by tweise** [comment]
test this please

**Comment by lukecwik** [comment]
Run Java Precommit

**Comment by lukecwik** [comment]
Run Java Precommit

**Pull request merged**